File: sdks/python/alancoin/realtime.py
"""Real-time WebSocket event streaming for Alancoin.

Provides an async client that connects to the Alancoin WebSocket endpoint
and delivers events via callbacks. Supports subscription filtering,
auto-reconnection with exponential backoff, and context manager usage.

Requires the ``websockets`` library::

    pip install alancoin[realtime]

Usage::

    import asyncio
    from alancoin import Alancoin
    from alancoin.realtime import EventType, RealtimeSubscription

    client = Alancoin("http://localhost:8080", api_key="ak_...")

    async def main():
        async with client.realtime(
            on_event=lambda e: print(e.type, e.data),
            subscription=RealtimeSubscription(
                event_types=[EventType.TRANSACTION, EventType.ESCROW_CREATED],
            ),
        ) as rt:
            await rt.wait_closed()

    asyncio.run(main())
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional, Any
from urllib.parse import quote

try:
    import websockets
    import websockets.exceptions
except ImportError as _err:
    raise ImportError(
        "The 'websockets' package is required for real-time streaming. "
        "Install it with: pip install alancoin[realtime]"
    ) from _err

logger = logging.getLogger("alancoin.realtime")


# ---------------------------------------------------------------------------
# Event types — mirrors internal/realtime/hub.go constants
# ---------------------------------------------------------------------------


class EventType:
    """Real-time event type constants matching the server protocol."""

    TRANSACTION = "transaction"
    AGENT_JOINED = "agent_joined"
    MILESTONE = "milestone"
    PRICE_ALERT = "price_alert"
    COALITION = "coalition"

    SESSION_CREATED = "session_created"
    SESSION_CLOSED = "session_closed"
    PROXY_SETTLEMENT = "proxy_settlement"

    ESCROW_CREATED = "escrow_created"
    ESCROW_DELIVERED = "escrow_delivered"
    ESCROW_CONFIRMED = "escrow_confirmed"
    ESCROW_DISPUTED = "escrow_disputed"

    STREAM_OPENED = "stream_opened"
    STREAM_CLOSED = "stream_closed"

    ALL: List[str] = [
        TRANSACTION, AGENT_JOINED, MILESTONE, PRICE_ALERT, COALITION,
        SESSION_CREATED, SESSION_CLOSED, PROXY_SETTLEMENT,
        ESCROW_CREATED, ESCROW_DELIVERED, ESCROW_CONFIRMED, ESCROW_DISPUTED,
        STREAM_OPENED, STREAM_CLOSED,
    ]


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass
class RealtimeEvent:
    """A real-time event received from the WebSocket stream."""

    type: str
    timestamp: str
    data: Dict[str, Any]

    @classmethod
    def from_dict(cls, raw: dict) -> RealtimeEvent:
        return cls(
            type=raw.get("type", ""),
            timestamp=raw.get("timestamp", ""),
            data=raw.get("data", {}),
        )


@dataclass
class RealtimeSubscription:
    """Subscription filter controlling which events are delivered.

    A default subscription (``all_events=True``) receives everything.
    Set specific filters to narrow the stream.
    """

    all_events: bool = True
    event_types: Optional[List[str]] = None
    agent_addrs: Optional[List[str]] = None
    service_types: Optional[List[str]] = None
    min_amount: Optional[float] = None

    def to_dict(self) -> dict:
        """Serialize to the JSON wire format matching the server protocol."""
        d: Dict[str, Any] = {"allEvents": self.all_events}
        if self.event_types:
            d["eventTypes"] = self.event_types
        if self.agent_addrs:
            d["agentAddrs"] = self.agent_addrs
        if self.service_types:
            d["serviceTypes"] = self.service_types
        if self.min_amount is not None:
            d["minAmount"] = self.min_amount
        return d


@dataclass
class RealtimeStats:
    """Connection statistics."""

    events_received: int = 0
    reconnects: int = 0


# ---------------------------------------------------------------------------
# Callback types
# ---------------------------------------------------------------------------

EventHandler = Callable[[RealtimeEvent], None]
ErrorHandler = Callable[[Exception], bool]  # return True to reconnect


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------


class RealtimeClient:
    """Async WebSocket client for real-time Alancoin events.

    Use as an async context manager::

        async with RealtimeClient(base_url, api_key, on_event=handler) as rt:
            await rt.wait_closed()
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        subscription: Optional[RealtimeSubscription] = None,
        on_event: Optional[EventHandler] = None,
        on_error: Optional[ErrorHandler] = None,
        reconnect_backoff: float = 1.0,
        reconnect_max: float = 30.0,
        ping_interval: float = 30.0,
    ) -> None:
        self._base_url = base_url
        self._api_key = api_key
        self._subscription = subscription or RealtimeSubscription()
        self._on_event = on_event
        self._on_error = on_error
        self._reconnect_backoff = reconnect_backoff
        self._reconnect_max = reconnect_max
        self._ping_interval = ping_interval

        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._read_task: Optional[asyncio.Task[None]] = None
        self._closed = False
        self._events_received = 0
        self._reconnects = 0
        self._current_backoff = reconnect_backoff

    # -- Public API ----------------------------------------------------------

    async def start(self) -> None:
        """Start the background read loop. Called automatically by __aenter__."""
        if self._read_task is not None:
            return
        self._read_task = asyncio.create_task(self._read_loop())

    async def subscribe(self, subscription: RealtimeSubscription) -> None:
        """Update the subscription filter on a live connection."""
        self._subscription = subscription
        if self._ws is not None:
            try:
                await self._ws.send(json.dumps(subscription.to_dict()))
            except websockets.exceptions.ConnectionClosed:
                pass  # will re-send on reconnect

    async def close(self) -> None:
        """Shut down the connection and background task."""
        if self._closed:
            return
        self._closed = True
        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
        if self._read_task is not None:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass

    async def wait_closed(self) -> None:
        """Block until the connection terminates (error or close)."""
        if self._read_task is not None:
            await self._read_task

    def stats(self) -> RealtimeStats:
        """Return connection statistics."""
        return RealtimeStats(
            events_received=self._events_received,
            reconnects=self._reconnects,
        )

    # -- Context manager -----------------------------------------------------

    async def __aenter__(self) -> RealtimeClient:
        await self.start()
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.close()

    # -- Internal ------------------------------------------------------------

    def _ws_url(self) -> str:
        u = self._base_url.rstrip("/")
        u = u.replace("https://", "wss://", 1).replace("http://", "ws://", 1)
        return f"{u}/ws?token={quote(self._api_key, safe='')}"

    async def _connect(self) -> websockets.WebSocketClientProtocol:
        url = self._ws_url()
        headers = {"X-API-Key": self._api_key}
        return await websockets.connect(
            url,
            additional_headers=headers,
            ping_interval=self._ping_interval,
            ping_timeout=self._ping_interval * 3,
            close_timeout=5,
        )

    async def _read_loop(self) -> None:
        while not self._closed:
            try:
                ws = await self._connect()
                self._ws = ws
                self._current_backoff = self._reconnect_backoff
                # Send initial subscription
                await ws.send(json.dumps(self._subscription.to_dict()))

                async for raw in ws:
                    if self._closed:
                        return
                    try:
                        data = json.loads(raw)
                        event = RealtimeEvent.from_dict(data)
                    except (json.JSONDecodeError, KeyError, TypeError):
                        continue  # skip malformed messages
                    self._events_received += 1
                    if self._on_event is not None:
                        try:
                            self._on_event(event)
                        except Exception:
                            logger.exception("on_event callback error")
                # Iterator exhausted — connection closed cleanly by server.
                # In production, websockets raises ConnectionClosed* exceptions
                # (caught below) rather than exhausting the iterator. A clean
                # iterator exit means the connection ended gracefully.
                return

            except asyncio.CancelledError:
                return
            except Exception as exc:
                if self._closed:
                    return
                logger.debug("WebSocket error: %s", exc)
                should_reconnect = True
                if self._on_error is not None:
                    try:
                        should_reconnect = self._on_error(exc)
                    except Exception:
                        logger.exception("on_error callback error")
                        should_reconnect = False
                if not should_reconnect:
                    return
                self._reconnects += 1
                await asyncio.sleep(self._current_backoff)
                self._current_backoff = min(
                    self._current_backoff * 2, self._reconnect_max
                )

File: sdks/go/realtime.go
package alancoin

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// Sentinel errors for realtime operations.
var (
	ErrConnClosed    = errors.New("alancoin: connection closed")
	ErrAlreadyClosed = errors.New("alancoin: already closed")
)

// EventHandler is called for each real-time event received from the server.
// It runs in a dedicated goroutine — blocking it delays subsequent events.
type EventHandler func(event RealtimeEvent)

// ErrorHandler is called when a WebSocket read or reconnect error occurs.
// Returning false stops automatic reconnection.
type ErrorHandler func(err error) (shouldReconnect bool)

// RealtimeConfig configures a [RealtimeClient].
type RealtimeConfig struct {
	// Subscription filters which events are delivered.
	// A zero-value Subscription (AllEvents = false, no filters) receives nothing.
	// Set AllEvents to true or populate EventTypes/AgentAddrs to receive events.
	Subscription RealtimeSubscription

	// OnEvent is called for each incoming event.
	OnEvent EventHandler

	// OnError is called on connection errors. Return true to reconnect.
	// If nil, the client reconnects automatically on transient errors.
	OnError ErrorHandler

	// ReconnectBackoff is the initial backoff delay between reconnect attempts.
	// Defaults to 1 second. Doubles on each attempt up to ReconnectMax.
	ReconnectBackoff time.Duration

	// ReconnectMax is the maximum backoff between reconnect attempts.
	// Defaults to 30 seconds.
	ReconnectMax time.Duration

	// PingInterval is how often a keepalive ping is sent to the server.
	// Defaults to 30 seconds. The server expects pings within 60 seconds.
	PingInterval time.Duration

	// ReadTimeout is how long to wait for a message before assuming the
	// connection is dead. Defaults to 90 seconds (3 missed pings).
	ReadTimeout time.Duration
}

func (cfg *RealtimeConfig) defaults() {
	if cfg.ReconnectBackoff == 0 {
		cfg.ReconnectBackoff = 1 * time.Second
	}
	if cfg.ReconnectMax == 0 {
		cfg.ReconnectMax = 30 * time.Second
	}
	if cfg.PingInterval == 0 {
		cfg.PingInterval = 30 * time.Second
	}
	if cfg.ReadTimeout == 0 {
		cfg.ReadTimeout = 90 * time.Second
	}
}

// RealtimeClient manages a persistent WebSocket connection to the Alancoin
// real-time event stream. It handles automatic reconnection, subscription
// management, and heartbeat pings.
//
// Usage:
//
//	rt, err := client.Realtime(ctx, alancoin.RealtimeConfig{
//	    Subscription: alancoin.RealtimeSubscription{AllEvents: true},
//	    OnEvent: func(e alancoin.RealtimeEvent) {
//	        fmt.Printf("%s: %v\n", e.Type, e.Data)
//	    },
//	})
//	if err != nil { ... }
//	defer rt.Close()
//
//	// Update subscription at any time:
//	rt.Subscribe(alancoin.RealtimeSubscription{
//	    EventTypes: []alancoin.EventType{alancoin.EventTransaction},
//	    MinAmount:  10.0,
//	})
type RealtimeClient struct {
	baseURL string
	apiKey  string
	http    *http.Client
	cfg     RealtimeConfig
	dialer  wsDialer

	conn   wsConn
	connMu sync.Mutex

	cancel context.CancelFunc
	done   chan struct{}

	closed atomic.Bool

	// Stats
	eventsReceived atomic.Int64
	reconnects     atomic.Int64
}

// wsConn abstracts a WebSocket connection for testing.
type wsConn interface {
	ReadMessage() (messageType int, p []byte, err error)
	WriteMessage(messageType int, data []byte) error
	WriteControl(messageType int, data []byte, deadline time.Time) error
	SetReadDeadline(t time.Time) error
	SetWriteDeadline(t time.Time) error
	SetPongHandler(h func(appData string) error)
	Close() error
}

// wsDialer abstracts WebSocket dialing for testing.
type wsDialer interface {
	dial(ctx context.Context, url string, header http.Header) (wsConn, *http.Response, error)
}

// defaultDialer uses gorilla/websocket (or falls back to a lightweight impl).
// Since the SDK has zero external dependencies (no gorilla/websocket), we use
// a pure-stdlib approach: we do not import gorilla/websocket. Instead, we
// provide a thin WebSocket layer using net/http + hijack, or the caller can
// inject a custom dialer. For production use, we recommend the gorilla-based
// dialer via WithWSDialer.
//
// The default implementation uses a raw HTTP-based approach that works for
// simple text-frame WebSocket communication.
type stdlibDialer struct {
	httpClient *http.Client
}

// Realtime opens a persistent WebSocket connection to the real-time event
// stream and returns a [RealtimeClient] that delivers events to the configured
// handler. The connection auto-reconnects on transient failures.
//
// The returned client must be closed when no longer needed.
func (c *Client) Realtime(ctx context.Context, cfg RealtimeConfig) (*RealtimeClient, error) {
	cfg.defaults()

	connCtx, cancel := context.WithCancel(ctx)

	var d wsDialer
	if c.wsDialer != nil {
		d = &wsDialerAdapter{d: c.wsDialer}
	}

	rt := &RealtimeClient{
		baseURL: c.baseURL,
		apiKey:  c.apiKey,
		http:    c.httpClient,
		cfg:     cfg,
		dialer:  d,
		cancel:  cancel,
		done:    make(chan struct{}),
	}

	// Attempt the initial connection.
	conn, err := rt.dial(connCtx)
	if err != nil {
		cancel()
		return nil, fmt.Errorf("alancoin: realtime connect: %w", err)
	}
	rt.connMu.Lock()
	rt.conn = conn
	rt.connMu.Unlock()

	// Send initial subscription.
	if err := rt.sendSubscription(conn, cfg.Subscription); err != nil {
		conn.Close()
		cancel()
		return nil, fmt.Errorf("alancoin: send subscription: %w", err)
	}

	// Start the read loop and keepalive pinger.
	go rt.runLoop(connCtx)

	return rt, nil
}

// Subscribe updates the event filter on a live connection. This takes effect
// immediately — the server will start applying the new filters to subsequent events.
func (rt *RealtimeClient) Subscribe(sub RealtimeSubscription) error {
	if rt.closed.Load() {
		return ErrAlreadyClosed
	}
	rt.cfg.Subscription = sub
	rt.connMu.Lock()
	conn := rt.conn
	rt.connMu.Unlock()
	if conn == nil {
		return ErrConnClosed
	}
	return rt.sendSubscription(conn, sub)
}

// Close terminates the WebSocket connection and stops all background goroutines.
func (rt *RealtimeClient) Close() error {
	if !rt.closed.CompareAndSwap(false, true) {
		return ErrAlreadyClosed
	}
	rt.cancel()
	<-rt.done // wait for runLoop to exit
	rt.connMu.Lock()
	conn := rt.conn
	rt.conn = nil
	rt.connMu.Unlock()
	if conn != nil {
		return conn.Close()
	}
	return nil
}

// Stats returns cumulative statistics for this connection.
func (rt *RealtimeClient) Stats() RealtimeStats {
	return RealtimeStats{
		EventsReceived: rt.eventsReceived.Load(),
		Reconnects:     rt.reconnects.Load(),
	}
}

// RealtimeStats holds connection-level statistics.
type RealtimeStats struct {
	EventsReceived int64
	Reconnects     int64
}

// runLoop is the main background loop that reads events and handles reconnection.
func (rt *RealtimeClient) runLoop(ctx context.Context) {
	defer close(rt.done)

	for {
		if err := ctx.Err(); err != nil {
			return
		}

		rt.connMu.Lock()
		conn := rt.conn
		rt.connMu.Unlock()
		if conn == nil {
			return
		}

		// Start a keepalive pinger in parallel.
		pingDone := make(chan struct{})
		go rt.pingLoop(ctx, conn, pingDone)

		// Read loop — blocks until error.
		readErr := rt.readLoop(ctx, conn)

		// Stop the pinger.
		close(pingDone)

		if ctx.Err() != nil {
			return // shutting down
		}

		// Connection lost — decide whether to reconnect.
		shouldReconnect := true
		if rt.cfg.OnError != nil {
			shouldReconnect = rt.cfg.OnError(readErr)
		}
		if !shouldReconnect || rt.closed.Load() {
			return
		}

		// Reconnect with exponential backoff.
		conn.Close()
		newConn, err := rt.reconnect(ctx)
		if err != nil {
			return // context cancelled or permanent failure
		}
		rt.connMu.Lock()
		rt.conn = newConn
		rt.connMu.Unlock()
		rt.reconnects.Add(1)
	}
}

// readLoop reads messages from the WebSocket until an error occurs.
func (rt *RealtimeClient) readLoop(ctx context.Context, conn wsConn) error {
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}

		_ = conn.SetReadDeadline(time.Now().Add(rt.cfg.ReadTimeout))
		_, msg, err := conn.ReadMessage()
		if err != nil {
			return err
		}

		var event RealtimeEvent
		if err := json.Unmarshal(msg, &event); err != nil {
			continue // skip malformed events
		}

		rt.eventsReceived.Add(1)
		if rt.cfg.OnEvent != nil {
			rt.cfg.OnEvent(event)
		}
	}
}

// pingLoop sends periodic pings to keep the connection alive.
func (rt *RealtimeClient) pingLoop(ctx context.Context, conn wsConn, done <-chan struct{}) {
	ticker := time.NewTicker(rt.cfg.PingInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-done:
			return
		case <-ticker.C:
			deadline := time.Now().Add(10 * time.Second)
			// WriteControl with PingMessage (opcode 9).
			if err := conn.WriteControl(9, nil, deadline); err != nil {
				return
			}
		}
	}
}

// reconnect attempts to re-establish the WebSocket connection with backoff.
func (rt *RealtimeClient) reconnect(ctx context.Context) (wsConn, error) {
	delay := rt.cfg.ReconnectBackoff
	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(delay):
		}

		conn, err := rt.dial(ctx)
		if err != nil {
			if delay < rt.cfg.ReconnectMax {
				delay *= 2
				if delay > rt.cfg.ReconnectMax {
					delay = rt.cfg.ReconnectMax
				}
			}
			continue
		}

		// Re-send subscription on the new connection.
		if err := rt.sendSubscription(conn, rt.cfg.Subscription); err != nil {
			conn.Close()
			continue
		}

		return conn, nil
	}
}

// dial establishes a new WebSocket connection.
func (rt *RealtimeClient) dial(ctx context.Context) (wsConn, error) {
	if rt.dialer != nil {
		header := http.Header{}
		if rt.apiKey != "" {
			header.Set("X-API-Key", rt.apiKey)
		}
		header.Set("User-Agent", userAgent)
		conn, _, err := rt.dialer.dial(ctx, rt.wsURL(), header)
		return conn, err
	}
	return nil, errors.New("alancoin: no WebSocket dialer configured; use WithWSDialer")
}

// wsURL converts the HTTP base URL to a WebSocket URL.
func (rt *RealtimeClient) wsURL() string {
	u := strings.TrimRight(rt.baseURL, "/")
	u = strings.Replace(u, "https://", "wss://", 1)
	u = strings.Replace(u, "http://", "ws://", 1)
	return u + "/ws"
}

// sendSubscription sends a subscription message over the WebSocket.
func (rt *RealtimeClient) sendSubscription(conn wsConn, sub RealtimeSubscription) error {
	data, err := json.Marshal(sub)
	if err != nil {
		return err
	}
	_ = conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
	return conn.WriteMessage(1, data) // TextMessage = 1
}

// dialer is the WebSocket dialer. Nil until set via WithWSDialer on the Client.
// We store it on the RealtimeClient to avoid adding a dependency.
var _ wsDialer = (*stdlibDialer)(nil)

// WithWSDialer configures a custom WebSocket dialer on the client. This is
// necessary because the SDK has no external WebSocket dependencies by default.
//
// For production use with gorilla/websocket:
//
//	import "github.com/gorilla/websocket"
//
//	dialer := alancoin.GorillaDialer(websocket.DefaultDialer)
//	client := alancoin.NewClient(url, alancoin.WithWSDialer(dialer))
func WithWSDialer(d WSDialer) Option {
	return func(c *Client) {
		c.wsDialer = d
	}
}

// WSDialer is the public interface for WebSocket dialers.
type WSDialer interface {
	DialContext(ctx context.Context, url string, header http.Header) (WSConn, *http.Response, error)
}

// WSConn is the public interface for WebSocket connections.
type WSConn interface {
	ReadMessage() (messageType int, p []byte, err error)
	WriteMessage(messageType int, data []byte) error
	WriteControl(messageType int, data []byte, deadline time.Time) error
	SetReadDeadline(t time.Time) error
	SetWriteDeadline(t time.Time) error
	SetPongHandler(h func(appData string) error)
	Close() error
}

// wsDialerAdapter wraps a public WSDialer into the internal interface.
type wsDialerAdapter struct {
	d WSDialer
}

func (a *wsDialerAdapter) dial(ctx context.Context, url string, header http.Header) (wsConn, *http.Response, error) {
	conn, resp, err := a.d.DialContext(ctx, url, header)
	if err != nil {
		return nil, resp, err
	}
	return conn, resp, nil
}

// stdlibDialer.dial is a stub — the stdlib approach requires gorilla or nhooyr
// to actually work. We provide the interface so users can inject their own.
func (d *stdlibDialer) dial(ctx context.Context, url string, header http.Header) (wsConn, *http.Response, error) {
	return nil, nil, errors.New("alancoin: stdlib WebSocket dialer not implemented; use WithWSDialer with gorilla/websocket")
}
